### PR TITLE
refactor(ingestion): unify memory task execution with heartbeat leases

### PR DESCRIPTION
**Pull request by jay77721**
### Summary / Motivation

Fixes #

Memory extraction shares the JetStream consumer and worker pool with document ingestion. Long-running or queued work can exceed the initial acknowledgement deadline, causing redelivery while the original owner is still active. Previously, a duplicate delivery could be acknowledged before the owner reached a durable outcome, and a retry after task-progress persistence failed could create duplicate extracted-memory documents.

This change keeps the broker lease alive from admission through settlement, preserves a single authoritative task identity, and makes memory retries idempotent.

### Key Changes

- Use the envelope `TaskMessage.TaskID` as the sole memory-task identity; remove duplicate identity fields from the payload.
- Introduce a reusable `Heartbeat` that renews the JetStream acknowledgement deadline for queued and running work.
- Keep duplicate deliveries unsettled and renew their lease while the owning worker is active; only the owner can acknowledge or negatively acknowledge the message.
- Stop lease renewal before settlement, then acknowledge or negatively acknowledge exactly once before releasing the in-process claim.
- Treat completed memory tasks as terminal on redelivery, propagate terminal progress-persistence failures as retryable, and generate stable extracted-memory document IDs across retries.
- Exclude runtime timestamp fallbacks from extracted-document fingerprints while retaining normalized explicit timestamps.

### Implementation Details

- Memory tasks receive their immutable ID from `TaskContext`, while the payload carries only extraction inputs.
- Admission starts a memory-task heartbeat before enqueueing so worker-pool backpressure cannot expire its lease.
- `Heartbeat.Stop` waits for in-flight renewal calls and safely handles a stop-before-start lifecycle.
- Extracted-memory IDs combine the memory ID, raw-message source ID, content, type, and only explicit normalized validity timestamps. Missing or malformed timestamps may still use a storage fallback, but cannot change the document ID on a later retry.

### How to Test

- `bash build.sh --test ./internal/ingestion/service ./internal/service`
- `bash build.sh --test -race ./internal/ingestion/service ./internal/service`
- Verify cross-attempt extracted-memory IDs remain equal when attempts use different times with omitted or malformed timestamps.

### CODE REVIEW

**Comment by jay77721** [comment]
## 📌 Staff Tech Lead Code Review: PR #19189

### Overview & Verdict
- **Verdict**: **Request Changes** (Quick fix required on memory chunk document ID stability)
- **Executive Summary**:
  1. The core architectural refactoring is outstanding: replacing `Ack` on duplicate deliveries with `InProgress` lease renewals eliminates silent task loss under worker crashes, and introducing the decoupled `Heartbeat` lifecycle manager brings solid stability to memory tasks.
  2. However, there is a critical bug in the claimed "retry idempotency": when LLM extraction does not produce an explicit `valid_at` (common for dialogue facts), `formatMemoryTime` falls back to the dynamic per-attempt `now`. This causes the computed document ID to drift between cross-second retries, resulting in duplicate chunks rather than upserts.
  3. Once the fallback timestamp is decoupled from the fingerprint computation and a cross-attempt unit test is added, this PR will be production-ready.

---

### 1. Problem Solved vs. Claimed

| Problem Claimed | Resolution Status | Technical Assessment |
|---|---|---|
| **Prevent Task Loss on Redelivery** | **Completely Solved** | Duplicate deliveries now call `handle.InProgress()` to renew broker leases while remaining unsettled, ensuring only the active worker settles the message. If the worker crashes, the broker redelivers without data loss. |
| **Memory Task Heartbeat & Claim Guard** | **Completely Solved** | Memory tasks now have admission-to-settlement lease renewals (queued in `taskChan` + worker execution), preventing NATS 5s timeout redelivery storms and worker pool starvation. |
| **Unified Task Identity** | **Completely Solved** | Task identity is anchored strictly to envelope `TaskMessage.TaskID` and immutable in `TaskContext.taskID`, eliminating payload discrepancy risks. |
| **Terminal Progress Error Propagation** | **Completely Solved** | DB progress update failures return `errMemoryTaskRetryable`, causing transient Nack retries rather than silent success. |
| **Stable Extracted Chunk ID Across Retries** | **Partially Solved (Has Loophole)** | Fingerprinting with `extractedMessageDocumentID` works when explicit timestamps exist, but breaks when `item.ValidAt` is omitted/empty. |

---

### 2. Critical Defect Analysis & Fix: Document ID Drift on Retry

#### The Root Cause
In `internal/service/memory_extractor.go`:
```go
validAt := formatMemoryTime(item.ValidAt, now)
invalidAt := ""
if strings.TrimSpace(item.InvalidAt) != "" {
    invalidAt = formatMemoryTime(item.InvalidAt, now)
}
...
"id": extractedMessageDocumentID(memoryID, sourceID, item.MessageType, item.Content, validAt, invalidAt),
```
When an extracted memory item has no explicit date (`item.ValidAt == ""`), `formatMemoryTime` falls back to `now.Format(memoryTimeLayout)`.
- If attempt 1 runs at $T_1$ (e.g. `10:00:00`), `validAt` is stamped with `10:00:00`.
- If saving to the chunk store succeeds but updating DB task progress fails, the task is retried.
- Attempt 2 runs at $T_2$ (e.g. `10:00:05`), `validAt` becomes `10:00:05`.
- SHA-256 fingerprinting produces a **different document ID**, causing the storage engine to insert an extra duplicate document instead of overwriting/upserting the previous one.

In addition, `TestBuildExtractedMessageUsesStableDocumentID` did not catch this because both `first` and `retry` were called with the exact same pinned `now` instance.

#### Recommended Remediation
Decouple the static fingerprinting timestamp from the runtime display fallback:

```go
// normalizeExplicitTime only normalizes explicit, parseable dates; returns empty string otherwise.
func normalizeExplicitTime(value string) string {
	v := strings.TrimSpace(value)
	if v == "" {
		return ""
	}
	for _, layout := range []string{time.RFC3339, "2006-01-02T15:04:05", "2006-01-02 15:04:05", "2006-01-02"} {
		if t, err := time.Parse(layout, v); err == nil {
			return t.Format(memoryTimeLayout)
		}
	}
	return ""
}

func buildExtractedMessage(messageID, sourceID int64, memoryID string, msg MemoryMessage, item extractedMemory, now time.Time) map[string]any {
	explicitValidAt := normalizeExplicitTime(item.ValidAt)
	explicitInvalidAt := normalizeExplicitTime(item.InvalidAt)

	storedValidAt := explicitValidAt
	if storedValidAt == "" {
		storedValidAt = now.Format(memoryTimeLayout)
	}
	var storedInvalidAt any
	if explicitInvalidAt != "" {
		storedInvalidAt = explicitInvalidAt
	}

	return map[string]any{
		// Fingerprint only over immutable content + explicit normalized timestamps
		"id":           extractedMessageDocumentID(memoryID, sourceID, item.MessageType, item.Content, explicitValidAt, explicitInvalidAt),
		"message_id":   messageID,
		"message_type": item.MessageType,
		"source_id":    sourceID,
		"agent_id":     msg.AgentID,
		"session_id":   msg.SessionID,
		"content":      item.Content,
		"valid_at":     storedValidAt,
		"invalid_at":   storedInvalidAt,
		"forget_at":    nil,
		"status":       true,
	}
}
```

#### Test Enhancement
Update `internal/service/memory_extractor_test.go`:
```go
func TestBuildExtractedMessageUsesStableDocumentIDAcrossAttempts(t *testing.T) {
	msg := MemoryMessage{UserID: "u1", AgentID: "a1", SessionID: "s1"}
	// ValidAt is omitted (empty) to simulate common LLM extracts
	item := extractedMemory{MessageType: "fact", Content: "likes coffee"}
	attempt1Time := time.Date(2026, 8, 20, 10, 5, 0, 0, time.Local)
	attempt2Time := time.Date(2026, 8, 20, 10, 5, 10, 0, time.Local) // 10s later retry

	first := buildExtractedMessage(8, 42, "mem-1", msg, item, attempt1Time)
	retry := buildExtractedMessage(9, 42, "mem-1", msg, item, attempt2Time)

	firstID, ok := first["id"].(string)
	if !ok || firstID == "" {
		t.Fatalf("first extracted message id = %#v, want stable non-empty string", first["id"])
	}
	if got, ok := retry["id"].(string); !ok || got != firstID {
		t.Fatalf("retry extracted message id = %#v, want %q across retry times", retry["id"], firstID)
	}
}
```

---

### 3. Compatibility & Deployment Considerations
- **Rolling Upgrade Risk**: In `queueMemoryTask`, removing `"id"` and `"task_id"` from the NATS payload means that if API servers are deployed before Ingestion workers in a rolling deployment, older workers trying to parse `taskCtx.MemoryPayload["id"]` will get an empty string.
  - *Recommendation*: In production deployments, ensure Ingestion workers are updated before API servers, or temporarily retain `"id": taskID` in the payload during the transition.
- **NATS `MaxAckPending`**: Duplicates staying in `InProgress` rather than being acknowledged means unacknowledged deliveries will occupy `MaxAckPending` until the owner finishes. Ensure consumer `MaxAckPending` headroom is adequately monitored under heavy backlog.

---

### 4. Summary
Except for the fallback timestamp fingerprinting defect, the design and test discipline in this PR are exemplary. Fixing the timestamp fingerprinting and adding the dynamic retry test will make this change completely bulletproof.

File: internal/service/memory_extractor.go [hunk 1]
```diff
@@ -221,11 +247,17 @@ func (s *MemoryMessageService) extractByLLM(ctx context.Context, mem *CreateMemo
...
+	invalidAt := ""
 	if strings.TrimSpace(item.InvalidAt) != "" {
 		invalidAt = formatMemoryTime(item.InvalidAt, now)
 	}
+	var storedInvalidAt any
+	if invalidAt != "" {
+		storedInvalidAt = invalidAt
+	}
 	return map[string]any{
+		"id":           extractedMessageDocumentID(memoryID, sourceID, item.MessageType, item.Content, validAt, invalidAt),
```
[the remark this answers is not in the document]
Comment (jay77721) [reply internal/service/memory_extractor.go]: Implemented in 83b70e720.

Fallback timestamps remain available for stored fields, but only explicit, successfully normalized valid_at and invalid_at values participate in the extracted-document fingerprint. Added cross-attempt coverage with a ten-second gap for both omitted valid_at and malformed timestamps, so retries reuse the same document ID.

Validated with:
- bash build.sh --test ./internal/ingestion/service ./internal/service
- bash build.sh --test -race ./internal/ingestion/service ./internal/service

**Comment by jay77721** [comment]
Follow-up to the review:

1. The retry-idempotency defect is fixed in 83b70e720. Fallback timestamps are excluded from the document fingerprint, and cross-attempt tests cover omitted and malformed timestamps.

2. We intentionally keep a single task identity path: TaskMessage.TaskID is authoritative and the payload does not retain id or task_id compatibility fields. The rollout requirement is to update all ingestors before publishers/API instances; new ingestors accept messages published by the preceding version.

3. MaxAckPending remains 131072. InProgress extends the existing delivery until its owner settles it; this is expected for long-running work. Pending depth and processing latency should be monitored in production, but no configuration change is included in this PR.

Validation:
- bash build.sh --test ./internal/ingestion/service ./internal/service
- bash build.sh --test -race ./internal/ingestion/service ./internal/service

**Comment by jay77721** [comment]
## Staff Tech Lead Review — PR #19189

### Verdict: **Request Changes** (one targeted fix required)

The PR makes a strong, well-tested set of improvements to the ingestion/memory-at-least-once path: a decoupled `Heartbeat` lifecycle manager, an in-process claim guard for memory tasks, the `errMemoryTaskRetryable` path for progress-persist failures, the `progress >= 1.0` redelivery short-circuit, and unified task identity via `TaskContext.taskID`. These are all solid. However, the central mechanism that is supposed to make retries idempotent — stable extracted-chunk document IDs — has a real gap for the common no-explicit-timestamp case, and the PR's own test for it does not exercise that path. That needs to be fixed before merge.

---

### 1. What the PR claims vs. what it delivers

| Claimed improvement | Assessment |
|---|---|
| Redelivering a duplicate no longer acks the only live copy; it renews the lease and stays unsettled until the owner settles. | **Correct.** Both the memory path (`processMessage` → `handle.InProgress()` on failed claim) and the ingestion path now renew the lease instead of acking. The owner is the sole settler. |
| Memory tasks get a broker lease from admission through queuing and execution. | **Correct.** `Heartbeat` is started in `processMessage` before the blocking send to `taskChan`, attached via `SetStopLease`, and stopped in the worker's `defer` *before* Ack/Nack. `TestProcessMessage_MemoryTaskHeartbeatsWhileQueued` covers the queued-but-unworked window that previously had no heartbeat. |
| The claim guard prevents double execution of a memory task within one ingestor. | **Correct.** `claimTask`/`releaseTask` now protect memory tasks; `TestProcessMessage_MemoryTaskRedeliveryRenewsLease` and `TestExecuteMemoryTask_ReleasesClaim` cover it. |
| Task identity is authoritative and never re-derived. | **Correct.** `TaskContext.ID()` returns only the envelope `taskID` captured at construction; the fallback chain in the old `executeMemoryTask` (`MemoryPayload["id"]` → `["task_id"]`) is gone. |
| Extracted-chunk IDs are stable across retries so reprocessing upserts, not duplicates. | **Partially solved — has a gap.** See §2. |
| Heartbeat `Stop-before-Start` lifecycle leak is fixed. | **Correct.** `Stop()` sets `stopped=true` without closing `stopCh` when no goroutine was started; a later `Start()` is a no-op. `TestHeartbeat_StopBeforeStartPreventsLaterStart` guards it. |

---

### 2. Required fix: document-ID drift when `valid_at` is absent

The retry-idempotency promise rests on `extractedMessageDocumentID` producing the same id on every attempt. It does not, for items without an explicit timestamp.

`internal/service/memory_extractor.go`:

```go
validAt := formatMemoryTime(item.ValidAt, now)          // now = memoryNow()
...
"id": extractedMessageDocumentID(memoryID, sourceID, item.MessageType, item.Content, validAt, invalidAt),
```

```go
func formatMemoryTime(value string, fallback time.Time) string {
	v := strings.TrimSpace(value)
	if v != "" { /* try a few layouts... */ }
	return fallback.Format(memoryTimeLayout)            // empty/unparseable → now
}
```

`memoryTimeLayout = "2006-01-02 15:04:05"` is second-precision, and `now` is re-evaluated per `saveExtractedToMemory` attempt. So for a dialogue-fact item with no `item.ValidAt` (the common case):

- Attempt 1 at `10:00:00` → `validAt = "2026-09-03 10:00:00"` → chunk id `mem-1_42_<hash1>`.
- Progress write fails → Nack → redelivery.
- Attempt 2 at `10:00:05` → `validAt = "2026-09-03 10:00:05"` → chunk id `mem-1_42_<hash2>` → **inserts a duplicate chunk instead of upserting.**

This directly undermines the PR's stated goal. The existing test `TestBuildExtractedMessageUsesStableDocumentID` does **not** catch it, because it calls both `buildExtractedMessage` invocations with the *same pinned `now`*. Stability across equal timestamps is not the property that matters; stability across *different* retry timestamps is.

**Fix:** fingerprint over content + only *explicitly* normalized timestamps; keep the dynamic wall-clock fallback solely for the stored `valid_at` field:

```go
explicitValidAt := normalizeExplicitTime(item.ValidAt)      // "" if absent/unparseable
explicitInvalidAt := normalizeExplicitTime(item.InvalidAt)

storedValidAt := explicitValidAt
if storedValidAt == "" {
    storedValidAt = now.Format(memoryTimeLayout)            // display only
}
var storedInvalidAt any
if explicitInvalidAt != "" { storedInvalidAt = explicitInvalidAt }

return map[string]any{
    "id":         extractedMessageDocumentID(memoryID, sourceID, item.MessageType, item.Content, explicitValidAt, explicitInvalidAt),
    "valid_at":   storedValidAt,
    "invalid_at": storedInvalidAt,
    ...
}
```

```go
func normalizeExplicitTime(value string) string {
	v := strings.TrimSpace(value)
	if v == "" { return "" }
	for _, layout := range []string{time.RFC3339, "2006-01-02T15:04:05", "2006-01-02 15:04:05", "2006-01-02"} {
		if t, err := time.Parse(layout, v); err == nil {
			return t.Format(memoryTimeLayout)
		}
	}
	return ""
}
```

And add the cross-attempt test (different `now` per attempt, empty `ValidAt`) to lock it in. This is a quick, low-risk change.

---

### 3. Supplementary observations (non-blocking)

**3a. Same-batch identical items collapse to one document id.** `extractedMessageDocumentID` keys on `(messageType, content, validAt, invalidAt)` only. If one source message yields two extracted items that are identical on those four fields, they map to the same id and the second overwrites the first *within the same batch*. The old random `messageID` avoided this. It's an edge case and arguably acceptable (duplicate facts are redundant), but if you want it bulletproof, add the batch index to the fingerprint: `fmt.Sprintf("%s_%d_%x_%d", memoryID, sourceID, fingerprint[:8], idx)`.

**3b. Rolling-upgrade contract risk in `queueMemoryTask`.** The payload no longer carries `id` / `task_id` / `task_type`; identity lives only on the envelope `TaskID`. If an API server (publisher) is redeployed *before* an ingestion worker (consumer) during a rolling upgrade, an old worker reading `MemoryPayload["id"]` gets `""`. The new code acks-and-skips empty ids in `processMessage`, so it degrades safely — but worth a deployment-order note (roll out ingestion workers first) or temporarily retaining `taskID` in the payload during the transition.

**3c. `MaxAckPending` headroom.** Duplicates now stay in `InProgress` instead of being acked immediately, so an unacknowledged redelivery occupies a `MaxAckPending` slot until the owner finishes. Under heavy backlog, monitor that headroom.

**3d. `embedAndSaveMessages` id-fallback branch is now effectively dead for the memory path.** Since `buildExtractedMessage` always sets `"id"`, the `if id, ok := message["id"].(string); !ok || id == ""` branch never fires for extracted messages. Not a bug (the raw-message path still relies on it), but a readability note.

---

### 4. Test coverage assessment

Good. The new tests target the hard-to-reach windows: queued-but-unworked heartbeat, lease-renewal on redelivery, claim release/re-claim, heartbeat `Stop-before-Stop` lifecycle, long-task heartbeat with the `wasSettledWithInProgress` race guard, and the `progress=1.0` short-circuit. The one gap is §2's cross-attempt ID-stability test with distinct timestamps — that's the test that would have caught the drift.

---

### 5. Bottom line

Fix the fingerprint so it does not depend on the dynamic `now` fallback (§2), add the cross-attempt test, and this PR is ready. Everything else — the `Heartbeat` redesign, the memory claim guard, the retryable-progress path, the redelivery short-circuit — is well-designed and well-tested.

**Comment by jay77721** [comment]
Thanks for the detailed follow-up review. I verified the findings against the current head, 83b70e720309a04fd0de859b810e6e0eb54108a5.

The required retry-idempotency issue is fixed in that commit. The stored fallback timestamp remains available for the document fields, but only explicitly supplied and successfully normalized timestamps participate in extracted document ID fingerprinting. The tests now cover retries at different times with both omitted and malformed timestamps.

Responses to the supplementary observations:

- 3a: Identical extracted facts in the same batch intentionally converge to one document ID. Adding a transient batch index would make retry identity depend on extraction order and could undermine idempotency. If identical items must remain distinct in the future, the discriminator should come from a stable source-level identity rather than the batch position.
- 3b: We intentionally do not retain payload compatibility fields. The rollout requirement is to update ingestion workers before publishers/API instances; the new worker accepts messages from the preceding producer version because the envelope TaskID is already present.
- 3c: MaxAckPending remains 131072. InProgress keeps the existing delivery unsettled until its owner finishes, so pending depth and processing latency should be monitored in production. No configuration change is needed for this PR.
- 3d: The ID fallback is still used by the raw-message path, even though extracted memory messages always provide an ID, so it remains intentional.

The targeted unit and race tests both pass. Please re-evaluate the required-change verdict against the current head, where the timestamp issue has been addressed.
